text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+implicit state to objects, since users must know what the default injections are
+ahead of time.
+
+By contrast, service getters (decorated with `@service`) clearly and explicitly
+state the dependencies of a class within the its definition. This is marginally
+more work, but the benefit of having an explicit dependencies list within each
+class has proven to be invaluable in practice since `inject()` was introduced.
+
+Glimmer components will only support explicit service declarations by default.
+However, the `injections` parameter is intentionally left as an an object with
+a single key (the owner) to allow us to change this in the future without
+breaking changes.
+
+### Lifecycle Hooks
+
+Legacy components have 13 major lifecycle hooks that run during 4 major phases
+of the component lifecycle, with some hooks running during multiple phases:
+
+1. **Initialization:**
+   * `init`
+
+2. **Initial Render:**
+   * `willInsertElement`
+   * `didInsertElement`,
+   * `didReceiveAttrs`
+   * `willRender`
+   * `didRender`.
+
+3. **Rerenders and Updates:**
+   * `didReceiveAttrs`
+   * `didUpdateAttrs`
+   * `willUpdate`
+   * `didUpdate`
+   * `willRender`
+   * `didRender`
+
+4. **Destruction:**
+   * `willDestroyElement`
+   * `didDestroyElement`
+   * `willDestroy`
+   * `destroy`
+
+
+Many of these hooks have overlapping or redundant functionality, and it's fairly
+confusing when to use which and what the differences are. We can simplify this
+cycle in a number of ways:
+
+* Hooks that run during multiple phases such as `didRender` and
+  `didRecieveAttrs` can be convenient at times, but also add mental overhead and
+  redundancy. We can remove these in favor of clearly delineated hooks which
+  only run during _one_ phase.
+
+* "Bookend" methods (`did*` and `will*`) can be confusing, since they require
+  some specific knowledge of what the "bookended" functionality is. For
+  instance, users almost _always_ want to use `didInsertElement` and
+  `willDestroyElement`, but the existence of their opposite bookends can make
+  this confusing. Additionally, the fact that `didReceiveAttrs` and
+  `didUpdateAttrs` do _not_ have opposing bookends is inconsistent with this
+  pattern.
+
+* Hooks that run immediately _before_ a render such as `didUpdateAttrs` are not
+  particularly useful, because the only functionality they have which
+  post-render methods do not is to modify a bound value without triggering a
+  rerender:
+
+  ```js
+  willUpdateAttrs() {
+    if (this.someBoolArg) {
+      this.set('someBoundValue', true);
+    }
+  }
+  ```
+
+  These types of values can almost always be better expressed using computed
+  properties, which will update when accessed during render, preventing
+  rerenders. Conversely, much imperative logic _must_ be done after a render
+  (manipulating an element directly, for instance), so post render hooks are
+  the most sensible.
+
+* The element based methods can be removed in favor of element modifiers,
```

Comment:
ya, this approach really does remove a large amount of complexity. Want to wait for x-child to be added/removed, simply use a modifier directly on it.